readme.md
[![Minimal node version](https://img.shields.io/static/v1?label=node&message=%3E=14.16&logo=node.js&color)](https://nodejs.org/about/releases/)
[![Minimal npm version](https://img.shields.io/static/v1?label=npm&message=%3E=6.14.12&logo=npm&color)](https://github.com/npm/cli/releases)
[![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg)](https://GitHub.com/Naereen/StrapDown.js/graphs/commit-activity)
[![Linux](https://svgshare.com/i/Zhy.svg)](https://svgshare.com/i/Zhy.svg)
[![macOS](https://svgshare.com/i/ZjP.svg)](https://svgshare.com/i/ZjP.svg)
[![Visual Studio Code](https://img.shields.io/badge/--007ACC?logo=visual%20studio%20code&logoColor=ffffff)](https://code.visualstudio.com/)

# BROZ

a simple, frameless browser for screenshots

It uses Cocoa/WebKit on macOS, gtk-webkit2 on Linux.

Inspire by [antfu broz](https://github.com/antfu/broz)

## Usage

### Install

Add .npmrc file in your env $HOME

> @stephen-shopopop:registry=https://npm.pkg.github.com

```
npx @stephen-shopopop/broz --url https://antfu.me

OR 

npm install -g  @stephen-shopopop/broz 
// launch
broz --url https://antfu.me
```

```
broz [options]

  --dir string
        path to serve (default ".")
  --url string
        instead of serving files, load this url
  --title string
        title of the webview window (default "broz")
  --width int
        width of the webview window (default 800)
  --height int
        height of the webview window (default 600)
```

## Distributing webview apps

On Linux you get a standalone executable. It will depend on GTK3 and GtkWebkit2,
so if you distribute your app in DEB or RPM format include those dependencies.
An application icon can be specified by providing a .desktop file.

```linux
sudo apt-get install libwebkit2gtk-4.0-dev
```

## Contributing

1. npm run lint - eslint test config
2. npm run release - Release library && publish
## Package maintenance

A modern cli tool that keeps your deps fresh

```bash
npx taze
```



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.24](https://github.com/stephen-shopopop/broz/compare/v0.0.23...v0.0.24) (2022-03-26)

### [0.0.23](https://github.com/stephen-shopopop/broz/compare/v0.0.22...v0.0.23) (2022-03-26)


### Bug Fixes

* **build:** name default build ([63388c3](https://github.com/stephen-shopopop/broz/commit/63388c3f4a94bcaa2c8a27aeb83b4b84b238c82a))

### [0.0.22](https://github.com/stephen-shopopop/broz/compare/v0.0.21...v0.0.22) (2022-03-26)

### [0.0.21](https://github.com/stephen-shopopop/broz/compare/v0.0.20...v0.0.21) (2022-03-26)


### Features

* **bundle:** Rename build ([134a547](https://github.com/stephen-shopopop/broz/commit/134a547a4e51880291e77ffeada24fa63c722d40))

### [0.0.20](https://github.com/stephen-shopopop/broz/compare/v0.0.19...v0.0.20) (2022-03-26)

### [0.0.19](https://github.com/stephen-shopopop/broz/compare/v0.0.18...v0.0.19) (2022-03-26)


### Bug Fixes

* **go:** default name ([45ae512](https://github.com/stephen-shopopop/broz/commit/45ae51276cc835c0a3c1e63ce5337e8ef04341dc))

### [0.0.18](https://github.com/stephen-shopopop/broz/compare/v0.0.17...v0.0.18) (2022-03-26)

### [0.0.17](https://github.com/stephen-shopopop/broz/compare/v0.0.16...v0.0.17) (2022-03-26)


### Features

* **main.go:** add automatic protocol ([359976d](https://github.com/stephen-shopopop/broz/commit/359976df9fabb5d10550edcd88a6cd5f34a05083))


### Bug Fixes

* package.json ([0d6d697](https://github.com/stephen-shopopop/broz/commit/0d6d6974be5ebf0987926e43896cc97400813e00))

### [0.0.16](https://github.com///compare/v0.0.15...v0.0.16) (2022-03-26)

### [0.0.15](https://github.com///compare/v0.0.14...v0.0.15) (2022-03-26)


### Bug Fixes

* fix npm ignore ([f6a780c](https://github.com///commit/f6a780c42a09dd6a0c47e522fbc0dc1e04aa2e9a))
* npmignore ([674f48a](https://github.com///commit/674f48ac5b709fbade26b87f6455a23b032ff8d8))

### [0.0.14](https://github.com///compare/v0.0.13...v0.0.14) (2022-03-26)

### [0.0.13](https://github.com///compare/v0.0.12...v0.0.13) (2022-03-26)


### Features

* builder ([b26d9b1](https://github.com///commit/b26d9b106d426c36b337265d623c6cc64d33e777))

### [0.0.12](https://github.com///compare/v0.0.11...v0.0.12) (2022-03-26)

### [0.0.11](https://github.com///compare/v0.0.8...v0.0.11) (2022-03-26)


### Features

* **bundle:** add build darwin ([a321c52](https://github.com///commit/a321c52f7bf8409c2428dc9545d29e6fb6f43718))
* **bundle:** bump versions ([9337ccb](https://github.com///commit/9337ccb5ebf419622e6720eebe6aca36151efbcf))
* **darwin:** add darwin version for go/nodejs/deno ([8b2ee26](https://github.com///commit/8b2ee268b08daa06938ace1783b62e743c846553))
* naming ([07925f7](https://github.com///commit/07925f7efe87cf5a2035784a8340db3429e1c6a1))


### Bug Fixes

* **bundle:** fix deno bundle ([c072a14](https://github.com///commit/c072a1409820c22bb87681eac3296e044d383f3f))
* **darwin:** fix bundle darwin ([0611a9d](https://github.com///commit/0611a9d55703010c35cbe2cbf653abe8deff7576))

### [0.0.10](https://github.com/stephendltg/go-webview/compare/v0.0.9...v0.0.10) (2021-08-16)

### [0.0.9](https://github.com/stephendltg/go-webview/compare/v0.0.8...v0.0.9) (2021-08-16)


### Features

* **darwin:** add darwin version for go/nodejs/deno ([8b2ee26](https://github.com/stephendltg/go-webview/commit/8b2ee268b08daa06938ace1783b62e743c846553))

### [0.0.8](https://github.com/stephendltg/go-webview/compare/v0.0.7...v0.0.8) (2021-08-04)

### [0.0.7](https://github.com/stephendltg/go-webview/compare/v0.0.6...v0.0.7) (2021-08-04)


### Features

* **bundle:** add win32 support ([d5b54f4](https://github.com/stephendltg/go-webview/commit/d5b54f442e07630b7d2bceb7e4f8ad8e074ce8ae))

### [0.0.6](https://github.com/stephendltg/go-webview/compare/v0.0.5...v0.0.6) (2021-08-04)

### [0.0.5](https://github.com/stephendltg/go-webview/compare/v0.0.4...v0.0.5) (2021-08-04)


### Features

* **deno:** add deno method ([59bfd99](https://github.com/stephendltg/go-webview/commit/59bfd993d5109499d5e42f253df579a814ac03ef))


### Bug Fixes

* **bundle:** change dir bin ([7b1a594](https://github.com/stephendltg/go-webview/commit/7b1a5945cc0e0ece7914303ec2ab4c2ffb93fdeb))
* **deno:** add dirname in path ([b82c379](https://github.com/stephendltg/go-webview/commit/b82c379fb78bc3e6bb8631de0441af8a66c691a7))

### [0.0.4](https://github.com/stephendltg/skeleton-go-webview/compare/v0.0.3...v0.0.4) (2021-08-03)

### [0.0.3](https://github.com/stephendltg/skeleton-go-webview/compare/v0.0.2...v0.0.3) (2021-08-03)

### Features

- **main.go:** add url or path for webview
  ([d6fe404](https://github.com/stephendltg/skeleton-go-webview/commit/d6fe40427886d0f8586ae47f0c62089a06da2dea))

### 0.0.2 (2021-08-03)

### Bug Fixes

- **git:** add hook git
  ([303a314](https://github.com/stephendltg/skeleton-go-webview/commit/303a314f79e2ddf7d6c484fe6faecfd9ecbd124d))


main.go
package main

import (
	"flag"
	"fmt"
	"net"
	"net/http"
	"os"
	"path/filepath"
	"strings"

	"github.com/webview/webview"
)

// Vars cli
var (
	title  = flag.String("title", "broz", "Webview title")
	debug  = flag.Bool("debug", false, "Mode debug")
	height = flag.Int("height", 600, "Webview height")
	width  = flag.Int("width", 800, "Webview width")
	url    = flag.String("url", "", "Navigate to this url")
	dir    = flag.String("dir", ".", "path to serve")
)

// Init
func init() {

	// Parse cli
	flag.Parse()

}

// Get port available
func getport() (int, net.Listener) {
	listener, err := net.Listen("tcp", ":0")
	if err != nil {
		panic(err)
	}

	port := listener.Addr().(*net.TCPAddr).Port

	return port, listener
}

// Handle http server
func handle(fs http.Handler) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		w.Header().Add("Referrer-Policy", "0")
		// Cache
		w.Header().Add("Surrogate-Control", "no-store")
		w.Header().Add("Cache-Control", "no-store, no-cache, must-revalidate, proxy-revalidate")
		w.Header().Add("Pragma", "no-cache")
		w.Header().Add("Expires", "0")
		// Security
		w.Header().Add("X-Content-Type-Options", "nosniff")
		w.Header().Add("X-DNS-Prefetch-Control", "off")
		w.Header().Add("X-Frame-Options", "DENY")
		w.Header().Add("Strict-Transport-Security", "5184000")
		w.Header().Add("X-Download-Options", "noopen")
		w.Header().Add("X-XSS-Protection", "1; mode=block")

		fs.ServeHTTP(w, r)
	}
}

// static server
func fileServer(listener net.Listener, path string) {
	http.Handle("/", handle(http.FileServer(http.Dir(path))))
	panic(http.Serve(listener, nil))
}

// Absolu path exec
func abspath() string {
	exe, err := os.Executable()
	if err != nil {
		panic(err)
	}
	exe, err = filepath.EvalSymlinks(exe)
	if err != nil {
		panic(err)
	}
	dir := filepath.Dir(exe)
	return dir
}

// Run webview
func runWebview(url string, title string, width int, height int, debug bool) {
	w := webview.New(debug)
	defer w.Destroy()
	w.SetTitle(title)
	w.SetSize(width, height, webview.HintNone)
	w.Navigate(url)
	// Func close webview for js
	w.Bind("close", func() {
		os.Exit(0)
	})
	w.Run()
}

func main() {

	port, listener := getport()

	var path string
	if strings.HasPrefix(*dir, "./") || strings.HasPrefix(*dir, "../") || strings.HasPrefix(*dir, ".\\") || strings.HasPrefix(*dir, "..\\") {
		path = filepath.Join(abspath(), *dir)
	} else {
		path = *dir
	}

	if *url == "" {
		go fileServer(listener, path)
		url := fmt.Sprintf("http://localhost:%d/", port)
		runWebview(url, *title, *width, *height, *debug)
	} else {
		var uri string
		if !strings.Contains(*url, "://") {
			uri = "http://" + *url
		} else {
			uri = *url
		}
		runWebview(uri, *title, *width, *height, *debug)
	}
}
